sklearn/utils/testing.py

```diff
@@ -372,14 +375,52 @@ def __exit__(self, *exc_info):
     assert_greater = _assert_greater
 
 
+def _sparse_dense_allclose(val1, val2, rtol=1e-7, atol=0):
+    """Check if two objects are close up to the preset tolerance.
+
+    The objects can be scalars, lists, tuples, ndarrays or sparse matrices.
+    """
+    if isinstance(val1, (int, float)) and isinstance(val2, (int, float)):
+        return np.allclose(float(val1), float(val2), rtol, atol)
+
+    if type(val1) is not type(val2):
+        return False
+
+    comparables = (float, list, tuple, np.ndarray, sp.spmatrix)
+
+    if not (isinstance(val1, comparables) or isinstance(val2, comparables)):
+        raise ValueError("The objects, %s and %s, are neither scalar nor "
+                         "array-like." % (val1, val2))
+
+    # list/tuple (or list/tuple of ndarrays/spmatrices)
+    if isinstance(val1, (tuple, list)):
+        if (len(val1) == 0) and (len(val2) == 0):
+            return True
+        if len(val1) != len(val2):
+            return False
+        while isinstance(val1[0], (tuple, list, np.ndarray, sp.spmatrix)):
```

Comment:
Yea sorry I'll correct it! (Initially I made it as a loop and later resorted to recursion)